File: backend/src/main/kotlin/com/nbweatherproject/backend/mapper/WeatherDataMapper.kt
package com.nbweatherproject.backend

import org.slf4j.LoggerFactory
import org.springframework.stereotype.Component
import java.time.LocalDate
import com.nbweatherproject.backend.OpenMeteoData

@Component
class WeatherDataMapper {
    companion object {
        private val logger = LoggerFactory.getLogger(WeatherDataMapper::class.java)
    }

    fun mapOpenMeteoToWeatherData(data: OpenMeteoData, queryId: Long): List<WeatherData> {
        logger.info("queryId in mapper: {}", queryId)
        return data.daily.time.indices.map { index ->
            WeatherData().apply {
                date = LocalDate.parse(data.daily.time[index])
                highTemp = data.daily.temperature2mMax[index]
                lowTemp = data.daily.temperature2mMin[index]
                this.queryId = queryId
                weather = mapWeatherCode(data.daily.weathercode[index])
                windSpeed = data.daily.windspeed10mMax[index]
            }
        }
    }

    private fun mapWeatherCode(code: Int): WeatherData.Weather {
        return when (code) {
            0 -> WeatherData.Weather.CLEAR
            1, 2 -> WeatherData.Weather.MAINLY_CLEAR
            3 -> WeatherData.Weather.PARTLY_CLOUDY
            45, 48 -> WeatherData.Weather.FOG
            51, 53, 55 -> WeatherData.Weather.DRIZZLE
            61, 63, 65 -> WeatherData.Weather.RAIN
            71, 73, 75 -> WeatherData.Weather.SNOW
            80, 81, 82 -> WeatherData.Weather.RAIN_SHOWERS
            85, 86 -> WeatherData.Weather.SNOW_SHOWERS
            95, 96, 99 -> WeatherData.Weather.THUNDERSTORM
            else -> WeatherData.Weather.UNKNOWN
        }
    }
}

File: backend-ts/src/utils/weatherDataMapper.ts
import { OpenMeteoData, WeatherData, Weather } from "../types";

export function mapOpenMeteoToWeatherData(
  data: OpenMeteoData,
  queryId: number
): WeatherData[] {
  console.log("queryId in mapper:", queryId);

  return data.daily.time.map((time, index) => ({
    date: new Date(time),
    highTemp: data.daily.temperature_2m_max[index],
    lowTemp: data.daily.temperature_2m_min[index],
    queryId: queryId,
    weather: mapWeatherCode(data.daily.weathercode[index]),
    windSpeed: data.daily.windspeed_10m_max[index],
  }));
}

function mapWeatherCode(code: number): Weather {
  switch (code) {
    case 0:
      return Weather.CLEAR;
    case 1:
    case 2:
      return Weather.MAINLY_CLEAR;
    case 3:
      return Weather.PARTLY_CLOUDY;
    case 45:
    case 48:
      return Weather.FOG;
    case 51:
    case 53:
    case 55:
      return Weather.DRIZZLE;
    case 61:
    case 63:
    case 65:
      return Weather.RAIN;
    case 71:
    case 73:
    case 75:
      return Weather.SNOW;
    case 80:
    case 81:
    case 82:
      return Weather.RAIN_SHOWERS;
    case 85:
    case 86:
      return Weather.SNOW_SHOWERS;
    case 95:
    case 96:
    case 99:
      return Weather.THUNDERSTORM;
    default:
      return Weather.UNKNOWN;
  }
}
